README
"echoping" is a small program to test (approximatively) performances
of a remote host by sending it TCP "echo" (or other protocol, like
HTTP) packets.

To install it, see the INSTALL file. Or type "./configure; make" if
you're in a hurry :-)

To use it, simply:

% echoping machine.somewhere.org

or use the options before the machine name (see the man page).

See the DETAILS file for various traps when benchmarking networks,
specially with this program

In any case, be polite: don't bother the remote host with many repeated 
requests, especially with large size. Ask for permission if you often
test hosts which aren't yours.

Current features:

- uses the protocols echo, discard, chargen, HTTP (with SSL/TLS if you
  wish), ICP or SMTP,
- uses UDP instead of TCP for the protocols which accept it (like echo), 
- can repeat the test and display various measures about it,
- supports IPv6 as well as IPv4,
- supports IDN (Unicode domain names),
- supports plugins written by you, to test even more protocols (see PLUGINS),

Examples of output:

(Simple test with 1000 bytes echo TCP packets)
% echoping -v -s 1000 mycisco
This is echoping, version 2.0.
Trying to connect to internet address 10.99.64.1 to transmit 256 bytes...
Connected...
Sent (1000 bytes)...
Checked
Elapsed time: 0.059597 seconds
%

(Repeated tests with average and median displayed.)
% echoping -n 10 mymachine
[...]
Minimum time: 0.170719 seconds (1500 bytes per sec.)
Maximum time: 0.211176 seconds (1212 bytes per sec.)
Average time: 0.184577 seconds (1387 bytes per sec.)
Median time: 0.181332 seconds (1412 bytes per sec.)

(Testing a Web server with an HTTP request for its home page.)
%  echoping -h / mywww
Elapsed time: 0.686792 seconds

(The exit status is set if there is any problem, so you can use echoping
to test repeatedly a Web server, to be sure it runs fine.)



-------------
The reference site for echoping is:

http://echoping.sourceforge.net/

The distribution is from:

http://sourceforge.net/project/showfiles.php?group_id=4581

The first version of echoping was done on October 1995.

Stephane Bortzmeyer <bortz@users.sourceforge.net>.

Starting in October 2011, echoping is LOOKING FOR A NEW MAINTAINER. I
do not have time to properly maintain it. Of course, since echoping is
free software, you do not need my authorization to adopt it but,
still, I promise I will do what I can to ease the transition.

Required skills: C, and network. echoping is quite simple but it has
many open bugs, including some that may have security consequences 
<https://sourceforge.net/tracker/?group_id=4581>.

--------------------- 
If you want to reward echoping and its original developer, you can
(but are not forced to do so, echoping is free software):

* Give money through Sourceforge's donation system
  <http://sourceforge.net/project/project_donations.php?group_id=4581>

* Use my Amazon's wish list
  <http://www.amazon.com/gp/registry/23ELBV1YZ93SC>

* Send me a postcard. I love postcards:
       Stephane Bortzmeyer
       127, rue Brancion
       75015 Paris
       France



$Id: README 431 2011-10-09 15:11:45Z bortz $





ChangeLog
6.1 (NOT YET)

	DO NOT FORGET maintenance branch
	
6.0.1 (2007-04-04)

	* Bug fixes only
	
6.0.0 (2007-03-26)

	* Plugins: several plugins (whois, postgresql, dns) are now
	shipped with echoping and you can write your own. All new
	protocols will now be supported via plugins. Consequences:
	   * you now need an Unix with a working dlopen(3)
	   * echoping now requires the popt parsing library
	* popt: echoping now depends on popt. One main change: options
	parsing is lightly different so some scripts *may* break. A
	typical example is that "old-school" Unix aggreation of arguments
	("echoping -v4 localhost") is no longer accepted (use "echoping -v
	-4 localhost").
        * Many bug fixes, see the list at Sourceforge
	
5.2.0 (2004-02-29)

	* IDN (Unicode domain names) support
	
5.1.0 (2003-11-09)

	* GNU TLS support (alternative to OpenSSL),
	* Small licence change,
	* Portability fixes (MacOS X)
	
5.0.1 (2002-11-04)

	* Small bug fixes,
	* Better error messages with invalid options

5.0.0 (2002-10-11)

	* Now uses getnameinfo/getaddrinfo. Two consequences:
	   - you can compile only on a recent Unix
	   - you have IPv6 support
        * With -v, now able to separate kernel (TCP) and application latency
	* Now displays the standard deviation for repeated tests. Warning:
	  this may break some scripts which parse the output of echoping.
	
4.2.0 (2001-03-16)
	
        * New -a and -A options to control the proxy/cache (disable caching).
	  Fixes bug/wish #232324
	* Support for microsecond wait (uses usleep)
	* New option (-p) to set socket priority (Linux only)
	* New option (-P) to set IP Type of Service octet
	* Bug fixes: #230384 (SSL records are different from text lines)

4.1.0 (2001-02-14)

	* Fix several SSL bugs (#132001, #130151)
	* Now requires an ANSI C compiler
	* Now uses only stdarg.h so should really compile everywhere

4.0.2 (2001-01-26)

	* Stupid bug introduced in 4.0.1 when compiling without SSL

4.0.1 (2001-01-26)

	* Fix SSL bug when using -n (#130078)

4.0.0 (2001-01-25)

	* SSL (Secure Sockets Layer) support
	* Now reading many bytes at a time (big performance improvment).
	  Uses the standard I/O library.
	* Better detection of the HTTP port
	* Bug fixes (AI/X, Solaris)
	
3.1.0 (never released)

	* SMTP support
	
3.0.1 (2000-07-06)

	* configure now tries to detect the name of the HTTP port
	* Fix an AIX bug (closes #108782)

3.0.0 (2000-06-30)

	* Autoconf/automake support
	* Fix a bug in ICP timeout
	  
2.2.1 
        * Portability fixes

2.2
        * Stable release, apart from ICP

2.1.0
	* Timeouts even on TCP connections.
	* Support of HTTP 1.1
	
2.0.1

	* Bug with getservbyname for HTTP connections fixed.
	
2.0.0 
	* HTTP support. For testing of Web servers and Web proxies.
	* Much better portability.
   

$Id: ChangeLog 404 2007-04-05 12:39:01Z bortz $






	

echoping.h
/* $Id: echoping.h 430 2011-01-20 16:10:23Z bortz $ */

/* Settings you should not change -- see below for changeable ones */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/* Settings you can change */

#define DEFLINE 256
#define UDPMAX 65535
/* Mostly for HTTP */
#define MAXTOREAD 150000
#ifdef SMTP
#define MAXSMTP 1024
#define MAXSMTPLINES 30
#endif

/* Probably too many inclusions but this is to keep 'gcc -Wall' happy... */
#include	<stdio.h>
#include        <stdlib.h>
#include	<sys/types.h>
#include	<netdb.h>
#include	<sys/socket.h>
#include        <netinet/tcp.h>
#include	<netinet/in.h>
#include	<arpa/inet.h>
#include        <stdarg.h>
#include        <sys/time.h>
#include        <errno.h>
#include        <unistd.h>
#include        <string.h>
#include        <signal.h>
#include        <math.h>
#include        <dlfcn.h>

/* popt library */
#include        <popt.h>

#ifdef OPENSSL
#include <openssl/crypto.h>
#include <openssl/x509.h>
#include <openssl/pem.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/rand.h>
#endif                          /* OpenSSL */

#ifdef GNUTLS
#include <gnutls/gnutls.h>
#endif

#ifdef LIBIDN
#include <stringprep.h>         /* stringprep_locale_to_utf8() */
#include <idna.h>               /* idna_to_ascii_from_utf8() */
#endif

#ifndef FALSE
#define FALSE 0
#endif
#ifndef TRUE
#define TRUE 1
#endif

#ifndef SIGALRM                 /* Linux... */
#define SIGALRM   14            /* alarm clock timeout */
#endif
#ifndef SIGINT                  /* Linux... */
#define SIGINT   2              /* interrupt, generated from terminal special char */
#endif

#ifndef INADDR_NONE             /* SunOS */
#define       INADDR_NONE (-1)
#endif

#ifndef SOL_IP
#define SOL_IP (getprotobyname("ip")->p_proto)
#endif

/* These entities should be in errno.h but some systems do not define
   them. */
#ifdef DECL_SYS_NERR
extern int      sys_nerr;
#endif

/* If we have it, use it */
#ifdef HAVE_SIGACTION
#define USE_SIGACTION 1
#endif
#ifdef HAVE_TOS
#define USE_TOS 1
#endif
#ifdef HAVE_SOCKET_PRIORITY
#define USE_PRIORITY 1
#endif


#ifndef HEADER_INCLUDED
typedef union _CHANNEL {
    FILE           *fs;
#ifdef OPENSSL
    SSL            *ssl;
#endif
#ifdef GNUTLS
    gnutls_session  tls;
#endif
} CHANNEL;

/* Do not use "short" for "boolean" because popt does not know this
   type. On a little-endian machine without alignment issues, it may
   work but not, for instance, on UltraSparc platforms. See for
   instance Debian bug #254322. */
typedef unsigned int boolean;

struct result {
    boolean         valid;
    struct timeval  timevalue;
};

boolean         timeout_flag;
struct echoping_struct {
    boolean         udp;        /* Use the UDP protocol (TCP is the default) */
    boolean         ttcp;
    boolean         only_ipv4;
    boolean         only_ipv6;
    boolean         verbose;
};
typedef struct echoping_struct echoping_options;
#ifndef IN_PLUGIN
/* The functions we will find in the plugin */
/* Initializes the plugin with its arguments. Returns the port name or number or NULL if the plugin wants to use the raw interface. */
typedef char   *(*init_f) (const int argc, const char **argv,
                           const echoping_options global_options);
init_f          plugin_init;
typedef void    (*start_f) (struct addrinfo *);
start_f         plugin_start;
typedef void    (*start_raw_f) ();
start_raw_f     plugin_raw_start;
typedef int     (*execute_f) ();
execute_f       plugin_execute;
typedef void    (*terminate_f) ();
terminate_f     plugin_terminate;
#endif

#endif

struct timeval  null_timeval;
struct timeval  max_timeval;

#define	ECHO_TCP_PORT	"echo"
#define	DISCARD_TCP_PORT	"discard"
#define	CHARACTER_GENERATOR_TCP_PORT	"chargen"
#define DEFAULT_HTTP_TCP_PORT "http"
#define DEFAULT_HTTPS_TCP_PORT "https"
#define DEFAULT_ICP_UDP_PORT "icp"

#ifdef HTTP
/* Use the old HTTP 1.0 protocol? If yes, set HTTP10 to 1*/
#undef HTTP10
#endif

#define	USE_ECHO	1
#define	USE_DISCARD	2
#define	USE_CHARGEN	3
#define	USE_HTTP	4
#define USE_ICP		5
#define USE_SMTP        6

#define CHARGENERATED " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefg";

char           *server;
#ifdef LIBIDN
char           *locale_server, *ace_server, *utf8_server;
#endif

/* My functions */

/* error.c */
void            usage();
void            err_sys(char *str, ...);
void            err_ret(char *str, ...);
void            err_quit(char *str, ...);
char           *sys_err_str();
/* writen.c */
int             writen();
/* readline.c */
int             readline();
#ifdef OPENSSL
int             SSL_readline();
#endif
#ifdef GNUTLS
ssize_t         TLS_readline();
#endif
/* util.c */
char           *random_string();
void            tvsub();
void            tvadd();
void            tvavg();
void            tvmin();
void            tvmax();
int             tvcmp();
void            tvstddev();
void            tvstddevavg();
double          tv2double();
struct timeval  double2tv();
/* http.c */
#ifdef HTTP
char           *make_http_sendline();
void            find_server_and_port();
/* This one has prototypes, for a very subtile compiler issue. */
int             read_from_server(CHANNEL fs, short ssl, boolean accept_redirects);
#endif

#ifdef ICP
#include "icp.h"
void           *make_icp_sendline();
int             recv_icp();
#ifndef HTTP
int             read_from_server();
#endif
#endif

#ifdef SMTP
int             smtp_read_response_from_server();
#endif

extern char    *progname;

extern boolean  timeout_flag;

#include "compilation.h"

#ifndef HEADER_INCLUDED
#define HEADER_INCLUDED
#endif
